design/service-apis.md

```diff
@@ -0,0 +1,384 @@
+values from this field for the Envoy Service resource. The specific Service field used for the values will depend on the
+network publishing type specified in the associated Contour. For example, if the network publishing type is "Internal",
+the operator will create an Envoy Service of type "ClusterIP" and set `clusterIPs: 1.2.3.4`.
+
+If `addresses` is unspecified, an IP address will be automatically assigned through existing Kubernetes mechanisms. If
+the specified address is invalid or unavailable, the operator will indicate this in Gateway status.
+
+The following is a Gateway example that specifies IP address "1.2.3.4". This address will be used for each Gateway
+listener:
+
+```yaml
+kind: Gateway
+apiVersion: networking.x-k8s.io/v1alpha1
+metadata:
+  name: example-gateway
+  namespace: projectcontour
+spec:
+  gatewayClassName: example-gatewayclass
+  addresses:
+    - type: IPAddress
+      value: 1.2.3.4
+```
+
+Although Gateway supports multiple address types, only "IPAddress" will be supported by the Contour Operator. If a named
+address is requested, the operator will surface the "Detached" listener status condition on the Gateway with the
+"UnsupportedAddress" reason.
+
+## Upgrades
+
+A GatewayClass of `controller: projectcontour.io/contour-operator` is equivalent to running the latest operator and
+Contour release. In this case, a Gateway of this class will be automatically upgraded when the operator is upgraded.
+Initially, `projectcontour.io/contour-operator` will be the only supported value. When support is added,
+`projectcontour.io/contour-operator/<version>` will be used to version GatewayClasses. Versioning a GatewayClass will
+consist of the following workflow:
+
+- The Platform Provider upgrades the operator image from x to y, triggering an upgrade of the operator. The
+  Contour CRD is updated if y includes API changes.
+- The status of `projectcontour.io/contour-operator/x` GatewayClasses are updated to "Waiting" and associated
+  Gateways are updated to "NoSuchGatewayClass". The details for accomplishing this step is TBD (see the Open Questions
+  section).
+- The operator will no longer manage Gateways of `projectcontour.io/contour-operator/x` GatewayClasses. These
+  Gateways will continue to function and traffic will continue to be proxied. Contour/Envoy will continue running x.
+- Optionally, the Platform Provider updates the Contour resource referenced by the GatewayClass, or creates and
+  associates a new Contour to the GatewayClass. For example, a new configuration field is exposed through the Contour
+  resource in y.
+- The Platform Provider updates Gateways by setting GatewayClasses to `projectcontour.io/contour-operator/y`. This
+  triggers the operator to upgrade Contour/Envoy to y and optionally update the configuration.
+  GatewayClass|Gateway|xRoute status conditions are updated accordingly.
+
+## Implementation Details
+
+### Controller
```

Comment:
I think we need to settle on nomenclature here, between how we refer to the thing that's in `proejctcontour/contour` at the moment, and how we refer to the Operator. I suggest that this should be `Operator` and the repo is the Contour operator, and we call `projectcontour/contour` the Contour controller. That lets us avoid things like "Contour proper", or other slightly tortured phrases.

Reply:
OK, I was trying to following the Service API terms, using "controller" instead of "operator".